### PR TITLE
Replace no data in variability with consensus

### PR DESCRIPTION
**Pull request by liviajakob**
When combining datasets together in the glambie algorithm it is essential to have an annual variability dataset. This is normally generated per data group individually, however there are data groups that don't have this information in certain regions, or where the annual variability does not span the full length. During the glambie workshop it was decided that this should be filled with the consensus variability from all datasets as defined in the config files. This requires to run the combination of annual variability into a consensus before running the algorithm in that region.

Note that depending on the run settings the consensus variability can still not end up spanning the whole length (e.g. in the case the run defines only one data_group), meaning that as a backup dataset the seasonal calibration dataset is used to fill the gaps.

This PR made changes to make that happen which includes:
- split up and refactoring of several functions (for variability and trends) which are used by the consensus variability as well as later in the algorithm: "_run_region_variability_for_one_source()" and "_run_region_trends_for_one_source()"
- ability to define time windows for a span of a timeseries, so that the consensus can be extended with the backup dataset if needed in "extend_annual_timeseries_if_shorter_than_time_window()". For this the ability to query min and max of all datasets combined in a data catalogue has also been added.
- New function which prepares the annual consensus variability "_prepare_consensus_variability_for_one_region()"
- Some quick fixes in the uncertainty calculation

### CODE REVIEW

File: glambie/processing/process_regional_results.py [hunk 1]
```diff
@@ -266,48 +355,23 @@ def _run_region_timeseries_one_source(
...
+        _run_region_variability_for_one_source(
+            data_catalogue_annual=data_catalogue_annual, seasonal_calibration_dataset=seasonal_calibration_dataset,
+            year_type=year_type, method_to_correct_seasonally=method_to_correct_seasonally,
+            data_group=data_group, dataset_names_where_split_at_gap=split_dataset_names_annual)
 
     # 2) LONGTERM TRENDS
     log.info("Using the following trend datasets for %s, %s : %s",
              data_group.long_name, region.long_name, str([d.user_group for d in data_catalogue_trends.datasets]))
-    log.info("Recalibrating with longterm trends within data group and region...")
-    # get catalogue with longerm datasets in same unit as calibration dataset (mwe)
```
Comment (TristanGoss) [inline glambie/processing/process_regional_results.py:290]: small typo here

File: glambie/processing/process_regional_results.py [hunk 2]
```diff
@@ -266,48 +355,23 @@ def _run_region_timeseries_one_source(
...
-        output_trend_date_range=min_max_time_window_for_longterm_trends)
 
-    # now treat case where trends are outside annual combined timeseries
-    annual_combined_full_ext = extend_annual_timeseries_if_outside_trends_period(
-        annual_timeseries=annual_combined,
-        data_catalogue_trends=data_catalogue_trends,
-        timeseries_for_extension=annual_backup_dataset.convert_timeseries_to_annual_trends(year_type=year_type))
-
-    # recalibrate
-    catalogue_calibrated_series = calibrate_timeseries_with_trends_catalogue(
```
Comment (TristanGoss) [inline glambie/processing/process_regional_results.py:305]: I can almost see the "committee" nature of the project in the repo structure; it would be good to take some time at the beginning of the next phase of GlaMBIE to refactor this after the ATBD is complete for this phase, I'm sure there's a lot of stuff here that looks more complicated than it is purely because of the order in which it was implemented.

File: glambie/processing/process_regional_results.py [hunk 3]
```diff
@@ -343,12 +407,140 @@ def _run_region_timeseries_one_source(
...
+    # now treat case where trends are outside annual combined timeseries
+    annual_combined_full_ext = extend_annual_timeseries_if_shorter_than_time_window(
+        annual_timeseries=annual_combined_dataset,
+        timeseries_for_extension=annual_backup_dataset.convert_timeseries_to_annual_trends(year_type=year_type),
+        desired_time_window=data_catalogue_trends_homogenized.get_time_span_of_datasets())
+
+    # recalibrate
+    catalogue_calibrated_series = calibrate_timeseries_with_trends_catalogue(
+        data_catalogue_trends_homogenized, annual_combined_full_ext)
+    # we dont remove trends as these are calibrated series with trends
```
Comment (TristanGoss) [inline glambie/processing/process_regional_results.py:543]: I don't understand this comment, can you re-write a bit?

File: glambie/processing/processing_helpers.py [hunk 4]
```diff
@@ -176,7 +176,7 @@ def convert_datasets_to_annual_trends(data_catalogue: DataCatalogue,
     year_type : YearType
         type of annual year, e.g hydrological or calendar
     method_to_correct_seasonally: SeasonalCorrectionMethod
-        method as to how long-term trends are correct when they don't start in the desired season, i.e. don't follow
+        method as to how annual trends are corrected when they don't start in the desired season, i.e. don't follow
         the desired annual grid defined with 'year_type'
     seasonal_calibration_dataset: Timeseries, by default None
         Timeseries dataset for seasonal calibration if trends are at annual resolution. Will be ignore if seasonal
```
Comment (TristanGoss) [inline glambie/processing/processing_helpers.py:182]: typo: ignore -> ignored

File: glambie/processing/processing_helpers.py [hunk 5]
```diff
@@ -407,59 +407,59 @@ def prepare_seasonal_calibration_dataset(region_config: RegionRunConfig,
...
-            # now update the annual timeseries object with the extended timeseries
-            annual_timeseries_copy.data.changes = np.array(df_merged["changes_x"])
-            annual_timeseries_copy.data.errors = np.array(df_merged["errors_x"])
-            annual_timeseries_copy.data.start_dates = np.array(df_merged["start_dates"])
-            annual_timeseries_copy.data.end_dates = np.array(df_merged["end_dates"])
-            return annual_timeseries_copy
+    if (desired_time_window[0] < min(annual_timeseries_copy.data.start_dates)) \
+            or (desired_time_window[1] > max(annual_timeseries_copy.data.end_dates)) \
+            or not annual_timeseries_copy.data.is_cumulative_valid():  # or the case where the timeseries has a gap
+        log.info("Extension of annual is performed, as the trends are longer than the annual timeseries")
```
Comment (TristanGoss) [inline glambie/processing/processing_helpers.py:439]: The grammar here is strange: can we avoid using Passive Voice in logging statements please? Ideally the statement should make it clear whether you are _about to_ do something or _have already done_ something.

File: glambie/processing/processing_helpers.py [hunk 6]
```diff
@@ -407,59 +407,59 @@ def prepare_seasonal_calibration_dataset(region_config: RegionRunConfig,
...
+
+        # Remove trend of timeseries for extension over the common time period
+        catalogue_dfs = [annual_timeseries_copy.data.as_dataframe(), timeseries_for_extension.data.as_dataframe()]
+        start_ref_period = np.max([df.start_dates.min() for df in catalogue_dfs])
+        end_ref_period = np.min([df.end_dates.max() for df in catalogue_dfs])
+        if not start_ref_period < end_ref_period:
+            warnings.warn("Warning when removing trends. No common period detected.")
```
Comment (TristanGoss) [inline glambie/processing/processing_helpers.py:446]: There's no need to include the word "Warning" in a warning statement.

**Review by TristanGoss — CHANGES_REQUESTED** [review]
Looks good, just some minor typos here.

I find it hard to follow the algorithmic stuff because this is clearly an "in progress" algorithm rather than the finished article, with a lot of flexibility built in that will be unnecessary later. Hopefully someone will get a chance to re-write it once the consortium settles on what the algorithm is.

File: glambie/processing/processing_helpers.py (hunk 6 above)
Comment (liviajakob) [reply glambie/processing/processing_helpers.py:446]: done

File: glambie/processing/processing_helpers.py (hunk 5 above)
Comment (liviajakob) [reply glambie/processing/processing_helpers.py:439]: Done

File: glambie/processing/process_regional_results.py (hunk 3 above)
Comment (liviajakob) [reply glambie/processing/process_regional_results.py:543]: done

File: glambie/processing/processing_helpers.py (hunk 4 above)
Comment (liviajakob) [reply glambie/processing/processing_helpers.py:182]: done

File: glambie/processing/process_regional_results.py (hunk 1 above)
Comment (liviajakob) [reply glambie/processing/process_regional_results.py:290]: done

**Comment by liviajakob** [comment]
Regarding your comment, yes it is not easy to implement a constantly changing algorithm and the codebase is somewhat reflecting that. The algorithm should be mostly set now for the first paper, but this will likely continue into the next round. There is already a vague list of potential requirements for the next phase.

**Review by TristanGoss — APPROVED** [review]
Thanks for addressing my comments, approved
